services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/model/MongoDbBasedCredentialsDao.java

```diff
@@ -0,0 +1,460 @@
+                new FindOptions(),
+                new UpdateOptions().setReturningNewDocument(true),
+                replaceCredentialsPromise);
+
+        return replaceCredentialsPromise.future()
+                .compose(result -> {
+                    if (result == null) {
+                        return MongoDbBasedDao.checkForVersionMismatchAndFail(
+                                deviceId,
+                                resourceVersion,
+                                getByDeviceId(tenantId, deviceId, resourceVersion, tracingContext));
+                    } else {
+                        LOG.debug("successfully updated credentials for device [tenant: {}, device-id: {}]",
+                                tenantId, deviceId);
+                        if (LOG.isTraceEnabled()) {
+                            LOG.trace("new document in DB:{}{}", System.lineSeparator(), result.encodePrettily());
+                        }
+                        span.log("successfully updated credentials");
+                        return Future.succeededFuture(result.getString(CredentialsDto.FIELD_VERSION));
+                    }
+                })
+                .recover(error -> {
+                    if (MongoDbBasedDao.isDuplicateKeyError(error)) {
+                        return Future.failedFuture(new ClientErrorException(
+                                tenantId,
+                                HttpURLConnection.HTTP_CONFLICT,
+                                "credentials (type, auth-id) must be unique for device"));
+                    } else {
+                        return Future.failedFuture(error);
+                    }
+                })
+                .recover(this::mapError)
+                .onFailure(error -> {
+                    LOG.debug("error updating credentials", error);
+                    TracingHelper.logError(span, "error updating credentials", error);
+                })
+                .onComplete(r -> span.finish());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Future<Void> delete(
+            final String tenantId,
+            final String deviceId,
+            final Optional<String> resourceVersion,
+            final SpanContext tracingContext) {
+
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(deviceId);
+        Objects.requireNonNull(resourceVersion);
+
+        final Span span = tracer.buildSpan("delete Credentials")
+                .addReference(References.CHILD_OF, tracingContext)
+                .withTag(TracingHelper.TAG_TENANT_ID, tenantId)
+                .withTag(TracingHelper.TAG_DEVICE_ID, deviceId)
+                .start();
+        resourceVersion.ifPresent(v -> span.setTag("resource_version", v));
+
+        final JsonObject removeCredentialsQuery = MongoDbDocumentBuilder.builder()
+                .withTenantId(tenantId)
+                .withDeviceId(deviceId)
+                .document();
```

Comment:
`resourceVersion` is unused while deleting the entry from the database.

Reply:
Hmm, we do remove credentials as part of deleting a device only, so in that case the credentials are being removed *unconditionally*. However, the DAO exposes generic CRUD operations and we should probably support optimistic locking in the same way here as we do elsewhere so I would tend to modifying the code to use the resource version instead of ignoring it. WDYT?

Reply:
Since DAO exposes generic CRUD operations, I think so too that we should make use of the resource version in the delete query to support optimistic locking.